src/shims/thread.rs

```diff
@@ -0,0 +1,138 @@
+use crate::*;
+use rustc_index::vec::Idx;
+use rustc_target::abi::LayoutOf;
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    fn pthread_create(
+        &mut self,
+        thread: OpTy<'tcx, Tag>,
+        _attr: OpTy<'tcx, Tag>,
+        start_routine: OpTy<'tcx, Tag>,
+        arg: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.tcx.sess.warn(
+            "thread support is experimental. \
+             For example, Miri does not detect data races yet.",
+        );
+
+        let new_thread_id = this.create_thread()?;
+        let old_thread_id = this.set_active_thread(new_thread_id)?;
+
+        let thread_info_place = this.deref_operand(thread)?;
+        let thread_info_type = thread.layout.ty
+            .builtin_deref(true)
+            .ok_or_else(|| err_ub_format!(
+                "wrong signature used for `pthread_create`: first argument must be a raw pointer."
+            ))?
+            .ty;
+        let thread_info_layout = this.layout_of(thread_info_type)?;
+        this.write_scalar(
+            Scalar::from_uint(new_thread_id.index() as u128, thread_info_layout.size),
```

Comment:
`as` casts should not be used in Miri ever, as they might lose information.
But also, `from_uint` takes `Into<u128>`, so no cast should be needed here.

Reply:
The cast seems to be necessary:
```
error[E0277]: the trait bound `u128: std::convert::From<usize>` is not satisfied
  --> src/shims/thread.rs:81:27
   |
81 |         this.write_scalar(Scalar::from_uint(thread_id.index(), dest.layout.size), dest)
   |                           ^^^^^^^^^^^^^^^^^ the trait `std::convert::From<usize>` is not implemented for `u128`
   |
   = help: the following implementations were found:
             <u128 as std::convert::From<bool>>
             <u128 as std::convert::From<std::net::Ipv6Addr>>
             <u128 as std::convert::From<std::num::NonZeroU128>>
             <u128 as std::convert::From<u16>>
           and 3 others
   = note: required because of the requirements on the impl of `std::convert::Into<u128>` for `usize`
   = note: required by `rustc_mir::interpret::Scalar::<Tag>::from_uint`
```
Am I missing something?

Reply:
Oh, it's a `usize`. Those are unfortunately not freely castable, yes. :/
But the fact that we should not ever use `as` remains. So here use `u64::try_from(thread_id.index()).unwrap()` instead. Or maybe don't use `usize` for `ThreadId`.